File: desktop/Phobri.Desktop.Tests/Models/ProtocolTests.cs
using Xunit;
using Phobri.Desktop.Models;
using System.Text.Json;

namespace Phobri.Desktop.Tests.Models;

public sealed class ProtocolTests
{
    [Fact]
    public void SmsMessage_SerializesAndDeserializes()
    {
        var original = new SmsMessage
        {
            Id = 12345,
            ThreadId = 678,
            Address = "+15551234567",
            ContactName = "Mom",
            Body = "Are you coming for dinner?",
            Date = 1714608000000,
            Type = SmsType.Inbox,
            Read = true
        };

        var json = JsonSerializer.Serialize(original, JsonContext.DefaultOptions);
        var deserialized = JsonSerializer.Deserialize<SmsMessage>(json, JsonContext.DefaultOptions);

        Assert.NotNull(deserialized);
        Assert.Equal(original.Id, deserialized.Id);
        Assert.Equal(original.Address, deserialized.Address);
        Assert.Equal(original.Body, deserialized.Body);
        Assert.Equal(original.Type, deserialized.Type);
    }

    [Fact]
    public void CallLogEntry_SerializesAndDeserializes()
    {
        var original = new CallLogEntry
        {
            Id = 789,
            Number = "+15551234567",
            ContactName = "Mom",
            Date = 1714608000000,
            Duration = 120,
            Type = CallType.Incoming
        };

        var json = JsonSerializer.Serialize(original, JsonContext.DefaultOptions);
        var deserialized = JsonSerializer.Deserialize<CallLogEntry>(json, JsonContext.DefaultOptions);

        Assert.NotNull(deserialized);
        Assert.Equal(original.Id, deserialized.Id);
        Assert.Equal(original.Number, deserialized.Number);
        Assert.Equal(original.Duration, deserialized.Duration);
        Assert.Equal(original.Type, deserialized.Type);
    }

    [Fact]
    public void ProtocolMessage_Request_Serializes()
    {
        var msg = ProtocolMessage.Request("sms.sync.request", new { after = 0, limit = 100 });

        var json = msg.ToJson();
        var deserialized = ProtocolMessage.FromJson(json);

        Assert.Equal(MessageType.Request, deserialized.Type);
        Assert.Equal("sms.sync.request", deserialized.Action);
        Assert.NotNull(deserialized.Id);
    }

    [Fact]
    public void ProtocolMessage_Push_Serializes()
    {
        var sms = new SmsMessage
        {
            Id = 1,
            Address = "+123",
            Body = "Hello",
            Date = 123456789,
            Type = SmsType.Inbox,
            Read = false
        };

        var msg = ProtocolMessage.Push("sms.new", sms);
        var json = msg.ToJson();

        Assert.Contains("\"sms.new\"", json);
        Assert.Contains("\"Hello\"", json);
    }

    [Fact]
    public void SmsMessage_DisplayName_FallsBackToAddress()
    {
        var withContact = new SmsMessage { Address = "+123", ContactName = "Bob" };
        Assert.Equal("Bob", withContact.DisplayName);

        var withoutContact = new SmsMessage { Address = "+123", ContactName = null };
        Assert.Equal("+123", withoutContact.DisplayName);
    }

    [Fact]
    public void CallLogEntry_DurationFormatted()
    {
        var call = new CallLogEntry { Duration = 125 };
        Assert.Equal("2:05", call.DurationFormatted);

        var zeroCall = new CallLogEntry { Duration = 0 };
        Assert.Equal("0:00", zeroCall.DurationFormatted);
    }
}

File: android/app/src/test/kotlin/com/phobri/android/model/ProtocolTests.kt
package com.phobri.android.model

import kotlinx.serialization.json.Json
import org.junit.Assert.*
import org.junit.Test

class ProtocolTests {

    private val json = Json { ignoreUnknownKeys = true }

    @Test
    fun `sms message serializes and deserializes`() {
        val original = SmsMessage(
            id = 12345,
            threadId = 678,
            address = "+15551234567",
            contactName = "Mom",
            body = "Are you coming for dinner?",
            date = 1714608000000,
            type = SmsType.INBOX,
            read = true
        )

        val jsonStr = json.encodeToString(SmsMessage.serializer(), original)
        val deserialized = json.decodeFromString<SmsMessage>(jsonStr)

        assertEquals(original.id, deserialized.id)
        assertEquals(original.address, deserialized.address)
        assertEquals(original.body, deserialized.body)
        assertEquals(original.type, deserialized.type)
    }

    @Test
    fun `call log entry serializes and deserializes`() {
        val original = CallLogEntry(
            id = 789,
            number = "+15551234567",
            contactName = "Mom",
            date = 1714608000000,
            duration = 120,
            type = CallType.INCOMING
        )

        val jsonStr = json.encodeToString(CallLogEntry.serializer(), original)
        val deserialized = json.decodeFromString<CallLogEntry>(jsonStr)

        assertEquals(original.id, deserialized.id)
        assertEquals(original.number, deserialized.number)
        assertEquals(original.duration, deserialized.duration)
        assertEquals(original.type, deserialized.type)
    }

    @Test
    fun `sms displayName falls back to address`() {
        val withContact = SmsMessage(address = "+123", contactName = "Bob")
        assertEquals("Bob", withContact.displayName)

        val withoutContact = SmsMessage(address = "+123", contactName = null)
        assertEquals("+123", withoutContact.displayName)
    }

    @Test
    fun `call duration formatted correctly`() {
        val call = CallLogEntry(duration = 125)
        assertEquals("2:05", call.durationFormatted)

        val zeroCall = CallLogEntry(duration = 0)
        assertEquals("0:00", zeroCall.durationFormatted)
    }

    @Test
    fun `protocol message request serializes`() {
        val msg = ProtocolMessage(
            type = MessageType.REQUEST,
            action = "sms.sync.request",
            id = "abc123"
        )

        val jsonStr = json.encodeToString(ProtocolMessage.serializer(), msg)
        val deserialized = json.decodeFromString<ProtocolMessage>(jsonStr)

        assertEquals(MessageType.REQUEST, deserialized.type)
        assertEquals("sms.sync.request", deserialized.action)
        assertEquals("abc123", deserialized.id)
    }
}
